impl/maven-core/src/main/java/org/apache/maven/project/DefaultProjectBuilder.java

```diff
@@ -700,11 +709,149 @@ private void initProject(MavenProject project, ModelBuilderResult result) {
+                    if (hasMainResources) {
+                        // Modular project with resources configured via <sources> - already added above
+                        if (!resources.isEmpty()
+                                && !hasOnlySuperPomDefaults(resources, baseDir, ProjectScope.MAIN.id())) {
+                            logger.warn("Legacy <resources> element is ignored because main resources are "
+                                    + "configured via <source><lang>resources</lang></source> in <sources>");
+                        }
+                        logger.debug(
+                                "Main resources configured via <sources> element, ignoring legacy <resources> element");
+                    } else {
+                        // Modular project without resources in <sources> - inject module-aware defaults
+                        if (!resources.isEmpty()
+                                && !hasOnlySuperPomDefaults(resources, baseDir, ProjectScope.MAIN.id())) {
+                            String message =
+                                    "Legacy <resources> element is ignored because modular sources are configured. "
+                                            + "Use <source><lang>resources</lang></source> in <sources> for custom resource paths.";
+                            logger.warn(message);
+                            result.getProblemCollector()
+                                    .reportProblem(new org.apache.maven.impl.model.DefaultModelProblem(
+                                            message,
+                                            Severity.WARNING,
+                                            Version.V41,
+                                            project.getModel().getDelegate(),
+                                            -1,
+                                            -1,
+                                            null));
+                        }
+                        logger.debug("Injecting module-aware main resource roots for {} modules", modules.size());
+                        for (String module : modules) {
+                            Path resourcePath = baseDir.resolve("src")
+                                    .resolve(module)
+                                    .resolve(ProjectScope.MAIN.id())
+                                    .resolve("resources");
+                            logger.debug("  - Adding main resource root: {} (module: {})", resourcePath, module);
+                            project.addSourceRoot(
+                                    createModularResourceRoot(baseDir, module, ProjectScope.MAIN, outputDirectory));
+                        }
+                    }
+                } else {
+                    // Classic (non-modular) project
+                    if (hasMainResources) {
+                        // Resources configured via <sources> - already added above
+                        if (!resources.isEmpty()
```

Comment:
Unnecessary check for `resources.isEmpty()` (see above comment).

Reply:
Done in 0c2451e7a30f8f427347b6c8b10bca3f551db02d. Removed the redundant isEmpty() checks and renamed the method to hasExplicitLegacyResources with inverted logic for clearer semantics (returns false when empty, true when explicit legacy resources exist).